content/technotes/2019-12-19-urls-checking/index.md

```diff
@@ -0,0 +1,456 @@
+link, we used it thanks to a code looping over all links. For the about
+50 links without replacement, we amended the posts by hand to make sure
+to take context into account (e.g. removing the link vs. removing the
+whole sentence presenting it).
+
+There were quite a few false positives i.e. actually valid URLs. This
+lead to some edits in
+[`crul::ok()`](https://docs.ropensci.org/crul/reference/ok.html) and the
+following wisdom:
+
+Sometimes you’ll get an error for the HEAD request but not the GET
+request.
+
+``` r
+# use get verb instead of head
+crul::ok("http://animalnexus.ca")
+```
+
+    ## [1] FALSE
+
+``` r
+crul::ok("http://animalnexus.ca", verb = "get")
+```
+
+    ## [1] TRUE
+
+Sometimes you’ll need an user-agend whose name does not contain crul.
+
+``` r
+# some urls will require a different useragent string
+# they probably regex the useragent string
+crul::ok("https://doi.org/10.1093/chemse/bjq042")
+```
+
+    ## GnuTLS recv error (-54): Error in the pull function.
+
+    ## [1] FALSE
+
+``` r
+crul::ok("https://doi.org/10.1093/chemse/bjq042", verb = "get", useragent = "foobar")
+```
+
+    ## [1] TRUE
+
+From short to long links
+------------------------
+
+We only identified short links using the bit.ly service. We found the
+corresponding link by running the function below. There were actually
+only 4 short links so that was quick.
+
+``` r
+get_long <- function(url){
+  crul::HttpClient$new(url)$get()$url
+}
+get_long("http://bit.ly/2JfrzmE")
+```
+
+    ## [1] "https://www.timeanddate.com/worldclock/fixedtime.html?msg=rOpenSci+Community+Call+on+Reproducible+Research+with+R&iso=20190730T09&p1=791&ah=1"
+
+http vs https
+-------------
+
+<!--html_preserve-->
+{{% tweet 1159812119099060224 %}} <!--/html_preserve-->
+
+We proceeded as previously when checking external links, except we used
+better settings for `crul::ok()`.
+
+``` r
+
+http <- dplyr::filter(all_urls, grepl("http\\:", url))
+http <- dplyr::mutate(http, https = sub("http\\:", "https:", url))
+unique_urls <- unique(http[, "https"])
+
+ok <- memoise::memoise(
+  ratelimitr::limit_rate(crul::ok, 
+                         ratelimitr::rate(1, 1)))
+
+get_ok <- function(url){
+  message(url)
+  ok(url, verb = "get", useragent = "Maëlle Salmon checking links")
+}
+
+unique_urls <- unique_urls %>%
+  dplyr::group_by(https) %>%
+  dplyr::summarise(ok = get_ok(https))
+
+http <- dplyr::left_join(http, unique_urls, by = "https")
+httpsok <- dplyr::filter(http, ok)
+
+modify_url <- function(index, df = httpsok) {
+  row <- df[index,]
+  readLines(row$filepath) %>%
+    stringr::str_replace_all(row$url, row$https) %>%
+    writeLines(row$filepath)
+  
+}
+
+purrr::walk(seq_len(nrow(httpsok)), modify_url)
+```
+
+You can [browse the related
+PR](https://github.com/ropensci/roweb2/pull/582/). Note that in the
+above, we could have used
+[`urltools`](https://github.com/Ironholds/urltools/) to parse URLs and
+extract their scheme.
```

Comment:
maybe in parentheses, since I imagine many people won't be familar with that term

> extract their scheme (http or https)